### PR TITLE
calculate significant numbers

### PR DESCRIPTION
**Pull request by Josephine-Marie**
add functions for calculating the upper and lower value limit for each choropleth object.

### CODE REVIEW

**Comment by Josephine-Marie** [comment]
Upper and lower value get generated from list of values. (from explicitly handing a list, not the "values" from choropleth.json)
amount of steps is now fixed to 6 (because of 6 legend entries) -> "num_colors" can get deleted from choropleth.json

@henhuy pls take a look and best case merge it

File: django_mapengine/choropleth.py [hunk 1]
```diff
@@ -63,18 +113,17 @@ def __calculate_steps(choropleth_config: dict, values: Optional[list] = None) ->
...
+                error_msg = "the given values are not valid or out of range"
                 raise ChoroplethError(error_msg)
-            min_value = min(values)
-            max_value = max(values)
-            num = choropleth_config["num_colors"]
-            step = (max_value - min_value) / (num - 1)
-            return [min_value + i * step for i in range(num - 1)] + [max_value]
+            min_value = __calculate_lower_limit(min(values))
+            max_value = __calculate_upper_limit(max(values))
+            step = (max_value - min_value) / 5
```
Comment (henhuy) [inline django_mapengine/choropleth.py:125]: Why did you remove dynamic number of colors and replaced it with fixed one?
Suggestion: Make 6 the default (as constant) but, allow more colors if needed (up to max of 9, as this is current limitation by colorbrewer)

File: django_mapengine/choropleth.py [hunk 2]
```diff
@@ -2,11 +2,61 @@
...
+    if mini == 0:
+        limit = mini
+        return limit
+    if mini < 1:
+        current = str(mini).split(".")[1]
+        limit = int(current[:1]) / 10
+    if mini > 1:
+        if isinstance(mini, float):
+            mini = str(mini).split(".")[0]
+        length = len(str(mini))
```
Comment (henhuy) [inline django_mapengine/choropleth.py:33]: I think those str-to-int-to-str conversions could be rewritten to mathematical functions, but I would be okay with this.

**Review by henhuy — CHANGES_REQUESTED** [review]
Thanks for implementing better min/max bounds!
one request: I would reset number of colors to dynamic amount as this could be useful if we want to show data with either less or even more colors.

File: django_mapengine/choropleth.py (hunk 2 above)
Comment (Josephine-Marie) [reply django_mapengine/choropleth.py]: yes, it is way better now! :+1:

**Pull request merged**